sympy/physics/optics/medium.py

```diff
@@ -69,50 +70,31 @@ class Medium(Symbol):
     """
 
     def __new__(cls, name, permittivity=None, permeability=None, n=None):
-        obj = super().__new__(cls, name)
-        obj._permittivity = sympify(permittivity)
-        obj._permeability = sympify(permeability)
-        obj._n = sympify(n)
+        if not isinstance(name, Str):
+            name = Str(name)
+
         if n is not None:
             if permittivity is not None and permeability is None:
-                obj._permeability = n**2/(c**2*obj._permittivity)
-            if permeability is not None and permittivity is None:
-                obj._permittivity = n**2/(c**2*obj._permeability)
-            if permittivity is not None and permittivity is not None:
-                if abs(n - c*sqrt(obj._permittivity*obj._permeability)) > 1e-6:
-                   raise ValueError("Values are not consistent.")
+                permeability = n**2/(c**2*permittivity)
```

Comment:
I think it is better to `_sympify` before any calculation, e.g. if a custom object has `_sympy_` defined, but not a `__pow__`/`__mul__`, it wouldn't work with this class.

Reply:
Is something like this done elsewhere in the codebase? I'm struggling to think of a situation where permittivity, permeability, or n would have `_sympy_` defined by not `__pow__/__mul__`, since permittivity, permeability, and n should be either scalars or expressions that can be evaluated to scalars. I don't have any problem making the change, but I'm just not sure if it's necessary.

Reply:
I would expect it to not have a sensible `__mul__` in ALL cases it has `_sympy_` defined, since `_sympy_` is a a method a user made class that is not a subclass of `Basic` can have to indicate how SymPy should work with it (or more precisely, how to convert that non-SymPy object to an equivalent SymPy object). It should also be possible to do this for user-classes by modifying `converter` in `core.sympify`. (which is also handled if that object is given to `_sympify`).

Reply:
OK I see. I'll make the change then. I assume there is not a significant performance overhead to sympifying everything before the calculations, since sympify will return any sympy object as it is. Am I right here?

Reply:
It's probably best just to move all `_sympify` calls to the top of the function like:
```
n = _sympify(n) if n is not None else n
```